sdk/core/azure-core/inc/azure/core/operation.hpp

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "azure/core/context.hpp"
+#include "azure/core/operation_status.hpp"
+#include "azure/core/response.hpp"
+
+#include <chrono>
+#include <string>
+#include <memory>
+
+namespace Azure { namespace Core {
+
+  //@param <T> The type of the final result of the long-running operation.
+  template <class T> class Operation {
+  private:
+    virtual std::unique_ptr<Http::RawResponse> PollInternal(Context& context) = 0;
+    virtual Response<T> PollUntilDoneInternal(std::chrono::milliseconds period, Context& context)
+        = 0;
+
+  protected:
+    OperationStatus m_status = OperationStatus::NotStarted;
+
+  public:
+    OperationStatus Status() const { return m_status; }
+
+    bool Done() const
+    {
+      return (
+          m_status == OperationStatus::Succeeded || m_status == OperationStatus::Cancelled
+          || m_status == OperationStatus::Failed);
+    }
+
+    bool HasValue() const { return (m_status == OperationStatus::Succeeded); }
+
+    virtual T Value() const = 0;
+
+    virtual std::string GetResumeToken() const = 0;
+
+    std::unique_ptr<Http::RawResponse> Poll()
+    {
+      Context context;
```

Comment:
Its not used by the test b/c we don't use a pipeline.  Jeffrey wants the pipelines to do the Cancel checks so that makes context uninteresting from the point of view of testing operation.

Reply:
I am a bit confused by this. Should the context value be set to something though? Maybe root context with infinite timeout?

My question about not passing an uninitialized local in the source code isn't really about the testing.

Reply:
yeah, this doesn't make sense here.
The Poll API should take the context as a parameter. It is an I/O operation.

Reply:
Customers call Poll() when they are doing the polling.  We do not want to force a customer to pass a context.  Its an additional concept and the customer may not care about it.

Reply:
> We do not want to force a customer to pass a context

We are doing that already. All Service APIs takes options which contains the context.
Let's follow the same pattern and add `PollOptions` as an optional parameter for calling `Poll()`.
We would have the context inside the PollOptions and only the cancellation-interested end-users could use this options structure.
It also offers the bag of options for any future setting we might need for polling(). 

@ahsonkhan the context is not `Uninitialized`, it is initialized based on its default constructor:
`Context() : m_contextSharedState(std::make_shared<ContextSharedState>()) {}`

We can always create and use a context like this. It basically becomes a root node. 
However, it is usually simpler to use the `Azure::Core::GetApplicationContext()` instead of creating a new context

Reply:
I switched to use the GetApplication(), it makes the code easier to read.  Thanks for the suggestion @vhvb1989 .